File: python/src/a4_prototype.py
"""
The intent is to specify the kinds of objects to create using a prototypical instance,
and create new objects by copying this prototype.

Key participants:
- Prototype: Declares an interface for cloning itself
- ConcretePrototype: Implements an operation for cloning itself
- Client: Creates a new object by asking a prototype to clone itself

Key benefits:
- Adds and removes products at runtime
- Specifies new objects by varying values
- Specifying new objects by varying structure
- Reduced subclassing

Real-world example: A graphic editor where users can clone existing shapes
(circle, rectangle) to create duplicates with the same properties, then modify them.
"""
from abc import ABC, abstractmethod
from copy import deepcopy
from dataclasses import dataclass
from typing import Any, Dict


# ----- Prototype -----
class Prototype(ABC):
    """Abstract Prototype"""

    @abstractmethod
    def clone(self) -> Any:
        """Create and return a copy of this object"""
        pass


# ----- Concrete Prototypes -----
@dataclass
class Circle(Prototype):
    radius: float = 10.0
    color: str = "red"
    x: float = 0.0
    y: float = 0.0

    def draw(self) -> None:
        print(f"Drawing Circle(radius={self.radius}, color={self.color}, "
              f"position=({self.x}, {self.y}))")

    def clone(self) -> "Circle":
        """Deep copy to create an independent clone"""
        return deepcopy(self)


@dataclass
class Rectangle(Prototype):
    width: float = 10.0
    height: float = 5.0
    color: str = "blue"
    x: float = 0.0
    y: float = 0.0

    def draw(self) -> None:
        print(f"Drawing Rectangle(width={self.width}, height={self.height}, "
              f"color={self.color}, position=({self.x}, {self.y}))")

    def clone(self) -> "Rectangle":
        return deepcopy(self)


# ----- Prototype Registry -----
class PrototypeRegistry:
    """
    Manages registered prototypes and can clone them on request.
    This is similar to the C++ version but uses a dictionary.
    """

    def __init__(self) -> None:
        self._prototypes: Dict[str, Prototype] = {}

    def register(self, name: str, prototype: Prototype) -> None:
        """Register a prototype for cloning later"""
        self._prototypes[name] = prototype
        print(f"[Registry] Registered prototype: {name}")

    def unregister(self, name: str) -> None:
        """Remove a prototype from the registry"""
        if name in self._prototypes:
            del self._prototypes[name]
            print(f"[Registry] Unregistered prototype: {name}")

    def clone(self, name: str) -> Prototype:
        """Create a new object by cloning the registered prototype"""
        if name not in self._prototypes:
            raise ValueError(f"Prototype '{name}' not found in registry")

        print(f"[Registry] Cloning prototype: {name}")
        return self._prototypes[name].clone()


# ----- Example usage -----
def prototype_example() -> None:
    """Demonstrates the Prototype pattern with shape cloning"""
    print("\n--- Prototype Pattern: Shape Cloning ---\n")

    registry = PrototypeRegistry()

    # Create and register template shapes
    red_circle = Circle(radius=10.0, color="red")
    blue_rect = Rectangle(width=4.0, height=2.0, color="blue")

    registry.register("red-circle", red_circle)
    registry.register("blue-rect", blue_rect)

    # Clone shapes and modify them
    print("\n1. Cloning red circle:")
    c1 = registry.clone("red-circle")
    c1.draw()

    print("\n2. Cloning another red circle and modifying:")
    c2 = registry.clone("red-circle")
    c2.radius = 20.0
    c2.x = 50.0
    c2.draw()

    print("\n3. Original is unchanged:")
    red_circle.draw()

    print("\n4. Cloning rectangle:")
    r1 = registry.clone("blue-rect")
    r1.draw()


if __name__ == "__main__":
    prototype_example()

File: rust/src/a4_prototype.rs
//! # Prototype Pattern
//!
//! The Prototype pattern specifies kinds of objects to create using a
//! prototypical instance, and creates new objects by cloning this prototype.
//!
//! ## Key Participants
//! - **Prototype**: Declares interface for cloning itself
//! - **ConcretePrototype**: Implements clone operation
//! - **Client**: Creates new objects by asking prototype to clone itself
//!
//! ## Benefits
//! - Adding/removing products at runtime
//! - Specifying new objects by varying values
//! - Reducing subclass creation
//! - Dynamic class loading

use std::collections::HashMap;

// Prototype trait
pub trait Prototype {
    fn clone_box(&self) -> Box<dyn Prototype>;
    fn get_name(&self) -> &str;
    fn draw(&self);
}

// Concrete Prototypes
#[derive(Clone)]
pub struct Circle {
    pub radius: f64,
    pub color: String,
}

impl Circle {
    pub fn new() -> Self {
        Self {
            radius: 10.0,
            color: "red".to_string(),
        }
    }

    pub fn set_radius(&mut self, radius: f64) {
        self.radius = radius;
    }

    pub fn set_color(&mut self, color: &str) {
        self.color = color.to_string();
    }
}

impl Prototype for Circle {
    fn clone_box(&self) -> Box<dyn Prototype> {
        Box::new(self.clone())
    }

    fn get_name(&self) -> &str {
        "Circle"
    }

    fn draw(&self) {
        println!("Drawing {} (radius={}, color={})", self.get_name(), self.radius, self.color);
    }
}

#[derive(Clone)]
pub struct Rectangle {
    pub width: f64,
    pub height: f64,
    pub color: String,
}

impl Rectangle {
    pub fn new() -> Self {
        Self {
            width: 10.0,
            height: 5.0,
            color: "blue".to_string(),
        }
    }

    pub fn set_size(&mut self, width: f64, height: f64) {
        self.width = width;
        self.height = height;
    }

    pub fn set_color(&mut self, color: &str) {
        self.color = color.to_string();
    }
}

impl Prototype for Rectangle {
    fn clone_box(&self) -> Box<dyn Prototype> {
        Box::new(self.clone())
    }

    fn get_name(&self) -> &str {
        "Rectangle"
    }

    fn draw(&self) {
        println!(
            "Drawing {} (width={}x{}, height={})",
            self.get_name(),
            self.width,
            self.height,
            self.color
        );
    }
}

// Prototype Registry
pub struct PrototypeRegistry {
    prototypes: HashMap<String, Box<dyn Prototype>>,
}

impl PrototypeRegistry {
    pub fn new() -> Self {
        Self {
            prototypes: HashMap::new(),
        }
    }

    pub fn register(&mut self, key: &str, prototype: Box<dyn Prototype>) {
        println!("Registering prototype: {}", key);
        self.prototypes.insert(key.to_string(), prototype);
    }

    pub fn create(&self, key: &str) -> Option<Box<dyn Prototype>> {
        self.prototypes.get(key).map(|p| p.clone_box())
    }
}

impl Default for PrototypeRegistry {
    fn default() -> Self {
        Self::new()
    }
}

/// Example demonstrating Prototype pattern
pub fn example() {
    println!("\n--- Prototype Pattern Example ---\n");

    let mut registry = PrototypeRegistry::new();

    // Set up prototype objects
    let mut red_circle = Circle::new();
    red_circle.set_radius(10.0);
    red_circle.set_color("red");

    let mut blue_rect = Rectangle::new();
    blue_rect.set_size(4.0, 2.0);
    blue_rect.set_color("blue");

    // Register prototypes
    registry.register("red-circle", Box::new(red_circle));
    registry.register("blue-rect", Box::new(blue_rect));

    // Clone and use prototypes
    println!("Creating objects from prototypes:");

    if let Some(r1) = registry.create("blue-rect") {
        r1.draw();
    }
}

// Helper trait for downcasting
pub trait AsAny {
    fn as_any(&self) -> &dyn std::any::Any;
}

impl<T: std::any::Any> AsAny for T {
    fn as_any(&self) -> &dyn std::any::Any {
        self
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_circle_clone() {
        let circle = Circle::new();
        let cloned = circle.clone_box();
        assert_eq!(cloned.get_name(), "Circle");
    }

    #[test]
    fn test_rectangle_clone() {
        let rect = Rectangle::new();
        let cloned = rect.clone_box();
        assert_eq!(cloned.get_name(), "Rectangle");
    }

    #[test]
    fn test_registry_register_and_retrieve() {
        let mut registry = PrototypeRegistry::new();
        registry.register("test-circle", Box::new(Circle::new()));

        let cloned = registry.create("test-circle");
        assert!(cloned.is_some());
        assert_eq!(cloned.unwrap().get_name(), "Circle");
    }

    #[test]
    fn test_registry_nonexistent_key() {
        let registry = PrototypeRegistry::new();
        let result = registry.create("nonexistent");
        assert!(result.is_none());
    }

    #[test]
    fn test_clone_independent_modification() {
        let mut original = Circle::new();
        original.set_radius(10.0);

        let cloned = original.clone_box();
        original.set_radius(20.0); // Modify original

        // Clone should have original radius (10.0)
        // This is a simple test - in real usage, we'd need downcasting
        cloned.draw();
    }
}
